Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp

```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+            }
+
+            GRefPtr<GVariant> printSettingsVariant(g_variant_lookup_value(options.get(), "settings", G_VARIANT_TYPE_VARDICT));
+            GRefPtr<GVariant> pageSetupVariant(g_variant_lookup_value(options.get(), "page-setup", G_VARIANT_TYPE_VARDICT));
+
+            priv->printSettings = gtk_print_settings_new_from_gvariant(printSettingsVariant.get());
+            priv->pageSetup = gtk_page_setup_new_from_gvariant(pageSetupVariant.get());
+
+            uint32_t token;
+            g_variant_lookup(options.get(), "token", "u", &token);
+
+            priv->preparePrintResponse = PreparePrintResponse(token);
+        }, g_object_ref(printOperation), g_object_unref);
+
+    GVariantBuilder options;
+    g_variant_builder_init(&options, G_VARIANT_TYPE_VARDICT);
+    g_variant_builder_add(&options, "{sv}", "handle_token", g_variant_new_string(token.ascii().data()));
+
+#if USE(SKIA)
+    GRefPtr<GVariant> portalVersion(g_dbus_proxy_get_cached_property(priv->portalProxy.get(), "version"));
+    if (portalVersion && g_variant_get_uint32(portalVersion.get()) >= 3)
+        g_variant_builder_add(&options, "{sv}", "supported_output_file_formats", g_variant_new_strv((const char* const[]) { "pdf" }, 1));
+#endif
+
+    const char* title = _("Print"); // FIXME
+    GRefPtr<GtkPageSetup> pageSetup = priv->pageSetup ? priv->pageSetup : adoptGRef(gtk_page_setup_new());
+    GRefPtr<GtkPrintSettings> printSettings(priv->printSettings ? gtk_print_settings_copy(priv->printSettings.get()) : gtk_print_settings_new());
+#if USE(SKIA)
+    gtk_print_settings_set(printSettings.get(), GTK_PRINT_SETTINGS_OUTPUT_FILE_FORMAT, "pdf");
+#endif
+
+    GRefPtr<GVariant> printSettingsVariant(gtk_print_settings_to_gvariant(printSettings.get()));
+    GRefPtr<GVariant> pageSetupVariant(gtk_page_setup_to_gvariant(pageSetup.get()));
+    GRefPtr<GVariant> arguments(g_variant_new("(ss@a{sv}@a{sv}a{sv})", "", title, printSettingsVariant.get(), pageSetupVariant.get(), &options));
+
+    g_dbus_proxy_call(priv->portalProxy.get(), "PreparePrint", arguments.get(), G_DBUS_CALL_FLAGS_NONE, -1, nullptr,
+        [](GObject* object, GAsyncResult* result, gpointer userData) {
+            auto printOperation = static_cast<WebKitPrintOperation*>(userData);
+            GDBusProxy* proxy = G_DBUS_PROXY(object);
+            GUniqueOutPtr<GError> error;
+            GRefPtr<GVariant> returnValue = adoptGRef(g_dbus_proxy_call_finish(proxy, result, &error.outPtr()));
+            if (error) {
+                WebKitPrintOperationPrivate* priv = printOperation->priv;
+                g_dbus_connection_signal_unsubscribe(g_dbus_proxy_get_connection(proxy), priv->signalId);
+                priv->signalId = 0;
+
+                printOperation->priv->preparePrintResponse = PreparePrintResponse(WEBKIT_PRINT_OPERATION_RESPONSE_CANCEL);
+            }
+        }, printOperation);
+
+    while (!priv->preparePrintResponse)
+        g_main_context_iteration(nullptr, false);
```

Comment:
```suggestion
    while (!priv->preparePrintResponse)
        g_main_context_iteration(nullptr, FALSE);
```

The need for nested main context iterations is unfortunate and dangerous, as discussed above. Is this one also unavoidable? I guess so....

Reply:
If that helps alleviate your concerns, I added code to push a separate GMainContext to the thread. This should at least prevent cross-context callbacks?

Reply:
Oh yes, using a separate main context is perfectly fine and safe. I didn't think of that. Excellent.